### DISCUSSION TITLE
My LDAP Search Filter seems broken with EMQX GUI but works on the command line with ldapsearch and apachedirectoryStudio

### DISCUSSION THREAD

**Discussion by kevdogg** [emqx/emqx#13710] [General]
So I'm kind of new to the EMQX program, but a step away from a novice with dealing with ldap.  I'm running my own openldap server (via osxia docker image) and I've been testing against this rudimentary database for several years.

The search I'm running is the following (manually through command line):
`ldapsearch -H ldap://openldap.domain.com:389 -x -D "cn=manager,dc=ldap,dc=domain,dc=com" -W -b "ou=users,dc=ldap,dc=domain,dc=com" -s sub -a always -z 1000 "(&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=domain,dc=com)))(uid=kevdog))" "userPassword"` 

This will return the following:
```
# extended LDIF
#
# LDAPv3
# base <ou=users,dc=ldap,dc=domain,dc=com> with scope subtree
# filter: (&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=domain,dc=com)))(uid=kevdog))
# requesting: userPassword
#

# kevdog, users, ldap.domain.com
dn: cn=kevdog,ou=users,dc=ldap,dc=domain,dc=com

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1
```

I've confirmed this search works within ApacheDirectoryStudio as well.

When running the search in the EQMX GUI I have the following options:
server: openldap.domain.com:389
Username: cn=manager,dc=ldap,dc=domain,dc=com
Password: ****
Base DN: ou=users,dc=ldap,dc=domain,dc=com
Password Authentication Mechanism: LDAP Bind Authentication
Bind Password: ${password}
(&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=domain,dc=com)))(uid=${username}))

When using MQTT Explorer to Connect (as this is my test client), I unfortunately get an error: Connection refused: Not Authorized

Digging into the eqmx logs (/var/log/emqx/emqx.log.1):
```
2024-08-27T17:32:56.648471-05:00 [error] clientid: mqtt-explorer-87b6208c, msg: filter_parse_failed, peername: 10.0.1.184:62330, username: kevdog, reason: {1,emqx_ldap_filter_parser,["syntax error before: ","equal"]}, filter: <<"(&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=domain,dc=com)))(uid=kevdog))">>
2024-08-27T17:32:56.648637-05:00 [error] tag: RESOURCE, clientid: mqtt-explorer-87b6208c, msg: unrecoverable_resource_error, peername: 10.0.1.184:62330, username: kevdog, reason: {1,emqx_ldap_filter_parser,["syntax error before: ","equal"]}, resource_id: <<"emqx_authn_ldap:1">>
2024-08-27T17:32:56.648738-05:00 [error] tag: AUTHN, clientid: mqtt-explorer-87b6208c, msg: ldap_query_failed, peername: 10.0.1.184:62330, username: kevdog, timeout: 5000, reason: {1,emqx_ldap_filter_parser,["syntax error before: ","equal"]}, resource: <<"emqx_authn_ldap:1">>, provider: emqx_authn_ldap_bind
2024-08-27T17:32:56.648843-05:00 [warning] tag: AUTHN, clientid: mqtt-explorer-87b6208c, msg: authentication_failure, peername: 10.0.1.184:62330, username: kevdog, reason: not_authorized
```

The relevant part of the is log is the following: ["syntax error before: ","equal"]}, filter: <<"(&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=domain,dc=com)))(uid=kevdog))"

I went ahead and played around with the filter a bit and if I ommited the memberof condition the search completed.

So I guess my question is -- what is wrong with my memberof attribute in the search function or is their something else I'm not finding?

**Comment by zmstone** [comment]
Hi @kevdogg 
EMQX LDAP integration does not support memberof search conditions so far.
You'll need to change to use a equality match condition instead.
Will bring this to the team for discussion.

BTW. We have been testing with openldap and microsoft AD, if you cold share some pointers how to test this using ApacheDirectoryStudio, it would be helpful. Thank you in advance.

**Comment by zmstone** [comment]
Hi @kevdogg github doesn't automatically link discussions when mentioned in PR.
It will be fixed in 5.8.1: https://github.com/emqx/emqx/pull/13864

**Comment by kevdogg** [comment]
Hey -- I've done some limited testing, but it seems like with 5.8.1, the member of search condition works.

Here is my search filter:
(&(&(|(objectclass=inetOrgPerson))(|(memberof=cn=mqtt,ou=groups,dc=ldap,dc=xxx,dc=com)))(uid=${username}))

The mqtt group has an object class of groupOfUniqueNames in which I have uniqueMembers defined.

Within the emqx I needed to add the user to the ACL. I wish there was a group placeholder but at looking at the ACL specs I don't think this is possible.

**Reply by zmstone** [reply]
placeholders are to render LDAP query requests,
if your idea is to extract information (in your case a `group` name) from query response, and turn it into a named value, then later use it to render a topic template for ACL rules --- this is not supported for LDAP authentication.

so far, only HTTP authentication can return extra data fields for this purpose: if `client_attrs : {key: value}` is provided in the response, `client_attrs.key` can be used in later stages.
However it's not a good example to use for ACL, because http auth response can directly return ACL like `acl: [{permission: allow, action: pub, topics : [...]}]` in the auth response.
